doc/js_tutorials/js_core/js_basic_ops/js_basic_ops.markdown

```diff
@@ -0,0 +1,380 @@
+--------- | ---------------
+CV_8U     | ucharAt
+CV_8S     | charAt
+CV_16U    | ushortAt
+CV_16S    | shortAt
+CV_32S    | intAt
+CV_32F    | floatAt
+CV_64F    | doubleAt
+
+@code{.js}
+let row = 3, col = 4;
+let src = cv.imread("canvasInput");
+let R = src.ucharAt(row, col * src.channels());
+let G = src.ucharAt(row, col * src.channels() + 1);
+let B = src.ucharAt(row, col * src.channels() + 2);
+let A = src.ucharAt(row, col * src.channels() + 3);
+@endcode
+
+@note  At manipulation is only for single channel access and the value can't be modified.
+
+**3. ptr**
+
+Mat Type  | Ptr Manipulation | JavaScript Typed Array
+--------  | ---------------  | ----------------------
+CV_8U     | ucharPtr         | Uint8Array
+CV_8S     | charPtr          | Int8Array
+CV_16U    | ushortPtr        | Uint16Array
+CV_16S    | shortPtr         | Int16Array
+CV_32S    | intPtr           | Int32Array
+CV_32F    | floatPtr         | Float32Array
+CV_64F    | doublePtr        | Float64Array
+
+@code{.js}
+let row = 3, col = 4;
+let src = cv.imread("canvasInput");
+let pixel = src.ucharPtr(row, col);
+let R = pixel[0];
+let G = pixel[1];
+let B = pixel[2];
+let A = pixel[3];
+@endcode
+
+mat.ucharPtr(k) get the k th row of the mat. mat.ucharPtr(i, j) get the i th row and the j th column of the mat.
+
+Image ROI
+---------
+
+Sometimes, you will have to play with certain region of images. For eye detection in images, first
+face detection is done all over the image and when face is obtained, we select the face region alone
+and search for eyes inside it instead of searching whole image. It improves accuracy (because eyes
+are always on faces) and performance (because we search for a small area)
+
+We use the function: **roi (rect)**
+@param rect    rectangle Region of Interest.
+
+Try it
+------
+
+Try this demo using the code above. Canvas elements named roiCanvasInput and roiCanvasOutput have been prepared. Choose an image and
+click `Try it` to see the result. You can change the code in the textbox to investigate more.
+
+\htmlonly
+<!DOCTYPE html>
+<head>
+<style>
+canvas {
+    border: 1px solid black;
+}
+.err {
+    color: red;
+}
+</style>
+</head>
+<body>
+<div id="roiCodeArea">
+<h2>Input your code</h2>
+<button id="roiTryIt" disabled="true" onclick="roiExecuteCode()">Try it</button><br>
+<textarea rows="7" cols="80" id="roiTestCode" spellcheck="false">
+let src = cv.imread("roiCanvasInput");
+let dst = new cv.Mat();
+// You can try more different parameters
+let rect = new cv.Rect(100, 100, 200, 200);
+dst = src.roi(rect);
+cv.imshow("roiCanvasOutput", dst);
+src.delete(); dst.delete();
+</textarea>
+<p class="err" id="roiErr"></p>
+</div>
+<div id="roiShowcase">
+    <div>
+        <canvas id="roiCanvasInput"></canvas>
+        <canvas id="roiCanvasOutput"></canvas>
+    </div>
+    <input type="file" id="roiInput" name="file" />
+</div>
+<script src="utils.js"></script>
+<script async src="opencv.js" id="opencvjs"></script>
```

Comment:
Loading from file (the `file://` protocol) has some access permission issues. It would be better to test with a local http server (e.g. node http-server) with `build_dir/doc/doxygen/html` as root. As it is similar to the web hosting environment.

When I test with a local web server, there is no such issue.

Reply:
Same problem via `http://`:
```
GET http://localhost:8000/doc/doxygen/html/de/d06/utils.js 
tutorial_js_basic_ops.html:269 GET http://localhost:8000/doc/doxygen/html/de/d06/opencv.js 
tutorial_js_basic_ops.html:268 GET http://localhost:8000/doc/doxygen/html/de/d06/utils.js 
tutorial_js_basic_ops.html:281 Uncaught ReferenceError: loadImageToCanvas is not defined
    at tutorial_js_basic_ops.html:281
(anonymous) @ tutorial_js_basic_ops.html:281
tutorial_js_basic_ops.html:355 Uncaught ReferenceError: loadImageToCanvas is not defined
    at tutorial_js_basic_ops.html:355
(anonymous) @ tutorial_js_basic_ops.html:355
tutorial_js_basic_ops.html:269 GET http://localhost:8000/doc/doxygen/html/de/d06/opencv.js 
```

Reply:
Thanks for trying the `http://`! It is wired to me why the URLs of `opencv.js` and `utils.js` have been modified to `de/06` etc., As you commented above, the URL is in the code embedded by `/htmlonly`. Will `doxgen` modify the URLs? Which build command you use? I use `python ./platforms/js/build_js.py <build_dir> --build_doc` which calls `make doxygen`. It works on my side.

Reply:
> de/06

It is part of URL for the tutorial page: `http://localhost:8000/doc/doxygen/html/de/d06/tutorial_js_basic_ops.html`
OpenCV uses this option for Doxygen:
```
CREATE_SUBDIRS         = YES
```

BTW, Any assets are located in the root directory, so `<iframe>` approach would work well.

Reply:
OK. I found the root cause. The `doxygen` I used is the default Ubuntu 16.04 one (1.8.11), however the opencv doc is generated by 1.8.12.

So can we assume loading the assets (example html) from root URL, e.g. `<iframe src="/js_basic_ops_roi.html"/>`?

Reply:
> from root URL

No. Documentation should work with URL sub-directories too. Sub-directories are used on docs.opencv.org by default.

I have tried `$relpath^` prefix but it doesn't work here. Perhaps we should use `../../` prefix as a workaround.

Reply:
> Perhaps we should use ../../ prefix as a workaround.

Thanks! I will implement that.